File: app/src/main/java/no/joharei/flixr/tools/ObservableCache.java
package no.joharei.flixr.tools;

import android.util.LruCache;

import javax.inject.Inject;
import javax.inject.Singleton;

import rx.Observable;

@Singleton
public class ObservableCache {
    private LruCache<String, Observable<?>> apiObservables;

    @Inject
    ObservableCache() {
        super();
        apiObservables = new LruCache<>(10);
    }

    public Observable<?> getCachedObservable(Observable<?> unPreparedObservable, Class<?> clazz, boolean cacheObservable, boolean useCache) {
        return getCachedObservable(unPreparedObservable, clazz, "", cacheObservable, useCache);
    }

    public Observable<?> getCachedObservable(Observable<?> unPreparedObservable, Class<?> clazz, String key, boolean cacheObservable, boolean useCache) {
        Observable<?> preparedObservable = null;

        if (useCache) {
            preparedObservable = apiObservables.get(clazz + key);
        }

        if (preparedObservable != null) {
            return preparedObservable;
        }

        preparedObservable = unPreparedObservable;

        if (cacheObservable) {
            preparedObservable = preparedObservable.cache();
            apiObservables.put(clazz + key, preparedObservable);
        }

        return preparedObservable;
    }

    public void clearCache() {
        apiObservables.evictAll();
    }
}

File: app/src/main/java/no/joharei/flixr/tools/ObservableCache.kt
package no.joharei.flixr.tools

import android.util.LruCache
import io.reactivex.Observable
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class ObservableCache
@Inject
internal constructor() {
    private val apiObservables = LruCache<String, Observable<*>>(10)

    fun <T> getCachedObservable(unPreparedObservable: Observable<T>, clazz: Class<T>, cacheObservable: Boolean, useCache: Boolean): Observable<T> {
        return getCachedObservable(unPreparedObservable, clazz, "", cacheObservable, useCache)
    }

    fun <T> getCachedObservable(unPreparedObservable: Observable<T>, clazz: Class<T>, key: String, cacheObservable: Boolean, useCache: Boolean): Observable<T> {
        var preparedObservable: Observable<T>? = null

        if (useCache) {
            preparedObservable = apiObservables.get(clazz.toString() + key) as? Observable<T>
        }

        if (preparedObservable != null) {
            return preparedObservable
        }

        preparedObservable = unPreparedObservable

        if (cacheObservable) {
            preparedObservable = preparedObservable.cache()
            apiObservables.put(clazz.toString() + key, preparedObservable)
        }

        return preparedObservable!!
    }

    fun clearCache() {
        apiObservables.evictAll()
    }
}
